R-packages/evalcast/R/calibration.R

```diff
@@ -1,70 +1,194 @@
-#' @importFrom magrittr %>%
-#' @importFrom rlang .data
-#' @importFrom purrr map_dfr map_dbl
-#' @importFrom dplyr group_by summarize mutate select
-compute_actual_vs_nominal_prob <- function(score_card) {
-  nominal_probs <- score_card$forecast_distribution[[1]]$probs
-  as.list(1:length(nominal_probs)) %>%
-    map_dfr(function(i) {
-      score_card %>%
-        group_by(.data$forecast_date) %>%
-        summarize(prop_below = mean(.data$actual < map_dbl(.data$forecast_distribution,
-                                                           ~ .x$quantiles[i]),
-                                    na.rm = TRUE))
-    }, .id = "nominal_id") %>%
-    mutate(nominal_prob = nominal_probs[as.integer(.data$nominal_id)]) %>%
-    select(-.data$nominal_id)
+
+
+check_valid_coverage_probs <- function(df, avg_vars) {
+  # The goal is to ensure that for a particular group, the quantiles at which
+  # forcasts are made are the same and reasonable
+  # Input is a grouped dataframe, with a column of "locations" that all need to
+  # comply. Rather than trying to locate some minimal set of "locations" that
+  # work, if there's a problem, we toss the whole group. The reason being,
+  # this likely reflects an error in the forecaster or the setup rather than
+  # something we should work around.
+  df <- df %>% filter(!is.na(.data$quantile)) # dump out any point forecasts
+  uq <- sort(unique(df$quantile))
+  n_uq <- length(uq)
+  # first, we check the whole group
+  valid <- n_uq %% 2 == 1 && is_symmetric(uq) # has median, and is symmetric
+  if (!valid) return(FALSE)
+  # now we check each "location"
+  df <- df %>% group_by(across(all_of(avg_vars))) %>%
+    summarise(valid = length(.data$quantile) == n_uq && # number of q's
+                # sorted + commensurate
+                all(find_quantile_match(.data$quantile, uq))) %>%
+    summarise(valid = all(valid))
+  return(df$valid)
+}
+
+#' Check if quantile forecasts are commensurate across groups
+#'
+#' @param cards a data frame containing at least the columns named
+#'   in `grp_vars`, `avg_vars` and a column called `quantile`
+#' @param grp_vars character vector of named columns in the score_card at which
+#'   average performance will be returned
+#' @param avg_vars character vector of named columns in the score_card over which
+#'   averages will be computed
+#' @param return_checks useful for debugging. returns a data frame indicating
+#'   which groups have valid or invalid avg_vars
+#'
+#' @return filtered cards based on which groups are completely valid 
+#'   (meaning "correct" quantiles are present for all `avg_vars` in the group)
+#' @export
+averaging_checks <- function(
+  cards, 
+  grp_vars = c("forecaster", "forecast_date", "ahead"),
+  avg_vars = c("geo_value"),
+  return_checks = FALSE) {
+    
+  gr <- cards %>% 
+    group_by(across(all_of(grp_vars)))
+  checks <- gr %>% group_keys()
+  checks$valid <- sapply(gr %>% group_split(), 
```

Comment:
Ran into an error while working on the vignette:
```
Error: Assigned data `sapply(...)` must be compatible with existing data.
x Existing data has 3 rows.
x Assigned data has 52 rows.
i Only vectors of size 1 are recycled.
```

I think you want `lapply` instead of `sapply`, but .data$valid is still a list.
```
    checks <- filter(checks, all(unlist(.data$valid))) %>% 
      select(-.data$valid)
```

Seems to work, but I wanted to confirm that that's the intended behavior.

Reply:
I think `sapply()` is ok here. It just reduces the list that `lapply()` would return to a vector. The issue is that `checks` has 3 rows, but `sapply(gr %>% group_split(),...)` is a length 52 vector. Can you show me how you're using the function?

Reply:
I'm just calling into it from the vignette:
```
plot_calibration(scorecards, 
                 grp_vars = "forecaster", 
                 avg_vars = c("geo_value", "forecast_date"))
```
Looking into it more, seems like `sapply` can work, I think the matrix is just transposed.
```
> foo = sapply(gr %>% group_split(), 
+        function(x) evalcast:::check_valid_coverage_probs(x, avg_vars))
> dim(foo)
[1] 52  3
> foo
      [,1] [,2] [,3]
 [1,] TRUE TRUE TRUE
 [2,] TRUE TRUE TRUE
 [3,] TRUE TRUE TRUE
 [4,] TRUE TRUE TRUE
 [5,] TRUE TRUE TRUE
...
```
I can add the transpose call.

Reply:
Hmmm... something seems strange about that. `check_valid_coverage_probs()` should (I think) not result in a matrix like that. So it must be a bug.

Reply:
Just wrapped the `sapply`, in a `t()` call. Leaving comment open for comment or to resolve.

Reply:
I found the bug. See line 22 in `calibration.R`